feat(curriculum): added the Tower of Hanoi Python project (#51347)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-recursion-by-solving-the-tower-of-hanoi-puzzle/64de6c8a5305d8173a3a9e09.md

  assert.match(code, /(?<=else\s*:\s+)print\s*\(\s*f('|")Moving\sdisk\s\{\s*rods\[\s*target\s*\]\[\s*-\s*1\s*\]\s*\}\sfrom\s\{\s*target\s*\}\sto\s\{\s*rod1\s*\}\1\s*\)/);
  assert.match(code, /rods\[\s*rod1\s*\]\.append\s*\(\s*rods\[\s*target\s*\]\.pop\s*\(\s*\)\s*\)/);
  } 
})
```

# --seed--

## --seed-contents--

```py
NUMBER_OF_DISKS = 3
number_of_moves = 2 ** NUMBER_OF_DISKS - 1
rods = {
    'A': list(range(NUMBER_OF_DISKS, 0, -1)),
    'B': [],
    'C': []
}
--fcc-editable-region--
def make_allowed_move(rod1, rod2):    
    forward = False
    if not rods[target]:
        forward = True
    elif rods[source] and rods[target] and rods[source][-1] < rods[target][-1]:
        forward = True              
    if forward:
        print(f'Moving disk {rods[source][-1]} from {source} to {target}')
        rods[target].append(rods[source].pop())
    else:
        print(f'Moving disk {rods[target][-1]} from {target} to {source}')

>         rods[source].append(rods[target].pop())

Author: With a couple of adjustments it works:
```py
def make_allowed_move(rod1, rod2):
    source, target = rod1, rod2
    if not rods[source]:
        source, target = target, source
    elif rods[target] and rods[source][-1] > rods[target][-1]:
        source, target = target, source

    print(f'Moving disk {rods[source][-1]} from {source} to {target}')
    rods[target].append(rods[source].pop())
    
    print(rods, '\n')
```
To be honest it seems less readable to me. And I don't think that having the default direction set from target to source is an issue in the current version. Anyway this is just my point of view and we can consider to proceed in this direction.

---
id: 64de6c8a5305d8173a3a9e09
title: Step 32
challengeType: 20
dashedName: step-32
---

# --description--

> `make_allowed_move` takes in `rod1` and `rod2` as parameters. You need a little refactoring here. Change each occurence of `source` into `rod1`.

Reviewer: ```suggestion
`make_allowed_move()` takes in `rod1` and `rod2` as parameters. You need a little refactoring here. Change every occurrence of `source` into `rod1`.
```

Reviewer: Was the term "refactoring" explained in a previous project? I would suggest adding more context about this in this step.

# --hints--

> You should change each occurence of `source` inside `make_allowed_move` into `rod1`.

Reviewer: ```suggestion
You should change each occurrence of `source` inside the `make_allowed_move()` function into `rod1`.
```

```js
({ test: () => {
  assert.match(code, /elif\s+rods\s*\[\s*rod1\s*\]\s+and\s+rods\s*\[\s*rod1\s*\]\[\s*-\s*1\s*\]\s*<\s*rods\s*\[\s*target\s*\]\[\s*-\s*1\s*\]\s*:/);
  assert.match(code, /if\s+forward(\s*==\s*True)?\s*:\s+print\s*\(\s*f('|")Moving\sdisk\s\{\s*rods\[\s*rod1\s*\]\[\s*-\s*1\s*\]\s*\}\sfrom\s\{\s*rod1\s*\}\sto\s\{\s*target\s*\}\2\s*\)\s+rods\[\s*target\s*\]\.append\s*\(\s*rods\[\s*rod1\s*\]\.pop\s*\(\s*\)\s*\)/);
  assert.match(code, /else\s*:\s+print\s*\(\s*f('|")Moving\sdisk\s\{\s*rods\[\s*target\s*\]\[\s*-\s*1\s*\]\s*\}\sfrom\s\{\s*target\s*\}\sto\s\{\s*rod1\s*\}\1\s*\)/);
  assert.match(code, /rods\[\s*rod1\s*\]\.append\s*\(\s*rods\[\s*target\s*\]\.pop\s*\(\s*\)\s*\)/);
  } 
})
```

# --seed--

## --seed-contents--

```py
NUMBER_OF_DISKS = 3

> number_of_moves = 2 ** NUMBER_OF_DISKS - 1

Reviewer: ```suggestion
number_of_moves = 2**NUMBER_OF_DISKS - 1
```

rods = {
    'A': list(range(NUMBER_OF_DISKS, 0, -1)),
    'B': [],
    'C': []
}
--fcc-editable-region--
def make_allowed_move(rod1, rod2):
    forward = False
    if not rods[target]:
        forward = True
    elif rods[source] and rods[source][-1] < rods[target][-1]:
        forward = True              
    if forward:
        print(f'Moving disk {rods[source][-1]} from {source} to {target}')
        rods[target].append(rods[source].pop())
    else:
        print(f'Moving disk {rods[target][-1]} from {target} to {source}')
        rods[source].append(rods[target].pop())
    

>     # display our progress

Reviewer: ```suggestion
    # Display our progress
```

    print(rods)
--fcc-editable-region--    

def move(n, source, auxiliary, target):

>     # display starting configuration

Reviewer: ```suggestion
    # Display starting configuration
```

    print(rods)
    for i in range(number_of_moves):
        remainder = (i + 1) % 3
        if remainder == 1:
            print(f'Move {i + 1} allowed between {source} and {target}')            
        elif remainder == 2:
            print(f'Move {i + 1} allowed between {source} and {auxiliary}')
        elif remainder == 0:
            print(f'Move {i + 1} allowed between {auxiliary} and {target}')

> # initiate call from source A to target C with auxiliary B

Reviewer: ```suggestion
# Initiate call from source A to target C with auxiliary B
```
